plugins/module_utils/k8s/runner.py

```diff
@@ -61,23 +63,30 @@ def run_module(module) -> None:
             if module.params.get("continue_on_error"):
                 result = {"error": "{0}".format(e)}
             else:
-                module.fail_json(msg=e)
+                module.fail_json(msg=to_native(e))
+
         if module.warnings:
             result["warnings"] = module.warnings
 
         results.append(result)
 
-    module.exit_json(**results)
+    module.exit_json(result=results)
 
 
 def perform_action(svc, definition: Dict, params: Dict) -> Dict:
     origin_name = definition["metadata"].get("name")
     namespace = definition["metadata"].get("namespace")
     label_selectors = params.get("label_selectors")
     state = params.get("state", None)
+    kind = definition.get("kind")
+    api_version = definition.get("apiVersion")
     result = {}
 
-    resource = svc.find_resource(definition)
+    if kind and kind.endswith("List"):
+        resource = svc.find_resource(kind, api_version, fail=False)
+    else:
+        resource = svc.find_resource(kind, api_version, fail=True)
+
```

Comment:
These changes shouldn't be necessary. If they are that means there's most likely a problem elsewhere.

Reply:
If I remove these lines here and keep only `resource = svc.find_resource(kind, api_version)`, this tasks fails https://github.com/ansible-collections/kubernetes.core/blob/2.x-refactor/molecule/default/roles/helm/tasks/test_crds.yml#L30 since the result won't contain `'Failed to find exact match for example.com/v1.Foo'` (it won't go here https://github.com/ansible-collections/kubernetes.core/blob/2.x-refactor/plugins/module_utils/k8s/service.py#L94 because fail is False).

Reply:
Ah, so what needs to happen then is just keep the `svc.find_resource(..., fail=True)` line. The old code conflates resource retrieval with building the list of definitions. We have separated those two things out in the refactor, so we should never have a List kind resource here. And if we did, none of the rest of this function is going to work anyways, so we can go ahead and fail here.